docs/development/design/build-images.rst

```diff
@@ -48,131 +45,242 @@ This produced issues to people pinning their images to any of these names becaus
+* Ubuntu 18 will still be available via ``stable`` and ``latest`` images, but not in new ones
+* a subset (not defined yet) of ``python``, ``nodejs``, ``rust`` and ``go`` versions on ``asdf`` are available to select
+
+.. note::
+
+   We are moving away from users specifying a particular Docker image.
+   With the new approach, users will specify the languages requirements they need,
+   and Read the Docs will decide if it will use a pre-built image or will spin up the base one and install these languages on the fly.
+
+   However, ``build.image`` will be still available for backward compatibility with ``stable``, ``latest`` and ``testing`` but won't support the new ``build.languages`` config.
+
+Note that knowing exactly what packages users are installing,
+could allow us to pre-build the most common combinations used images: ``ubuntu20-py39+node14``.
 
-.. admonition:: Implementation
 
-   We talked about using a ``Dockerfile.custom`` and build it on every build.
-   However, at this point it requires extra work to change our build pipeline.
-   We decided to install OS packages from the application itself for now using
-   Docker API to call ``docker exec`` as ``root`` user.
+Time required to install languages at build time
+------------------------------------------------
 
-   This reduces the amount of work required but also allows us to add this feature
-   to our current existing images (they require a rebuild to add ``nodenv`` and ``rustup``)
+Testings using ``time`` command in ASG instances to install extra languages took these "real" times:
 
+* ``build-default``
 
-Updating versions over time
----------------------------
+  * python 3.9.6: 2m21.331s
+  * mambaforge 4.10.1: 0m26.291s
+  * miniconda3 4.7.12: 0m9.955s
+  * nodejs 14.17.5: 0m5.603s
+  * rust 1.54.0: 0m13.587s
+  * golang 1.17: 1m30.428s
+
+* ``build-large``
+
+  * python 3.9.6: 2m33.688s
+  * mambaforge 4.10.1: 0m28.781s
+  * miniconda3 4.7.12: 0m10.551s
+  * nodejs 14.17.5: 0m6.136s
+  * rust 1.54.0: 0m14.716s
+  * golang 1.17: 1m36.470s
+
+Note that the only one that required compilation was Python.
+All the others, spent 100% of its time downloading the binary.
+These download times are *way better from EU* with a home internet connection.
+
+In the worst scenario: "none of the specified language version has a pre-built image",
+the build will require ~5 minutes to install all the language requirements.
+By providing *only* pre-built images with the Python version (that's the most time consuming),
+build times will only require ~2 minutes to install the others.
+However, requiring one version of each language is not a common case.
+
+Cache language binaries on S3
+-----------------------------
+
+``asdf`` scripts can be altered to download the ``.tar.gz`` dist files from a different mirror than the official one.
+Read the Docs can make usage of this to create a mirror hosted locally on S3 to get faster download speeds.
```

Comment:
Also it should be more reliable, though we pay the bandwidth fees. We should confirm we can do something that keeps the costs low, while also not hammering external mirrors.

Reply:
How we can confirm this upfront?

Reply:
Research if we're billed for S3 traffic internally

Reply:
It seems that we are not charged by EC2 <-> S3 traffic within the same region.

> You pay for all bandwidth into and out of Amazon S3, except for the following:
> ...
> Data transferred from an Amazon S3 bucket to any AWS service(s) within the
  same AWS Region as the S3 bucket (including to a different account in the same
  AWS Region).
> ...

(from "Data transfer" tab in https://aws.amazon.com/s3/pricing/)

----

Also, I checked using the AWS Cost Explorer I haven't find any charge for internal traffic.

https://aws.amazon.com/blogs/mt/using-aws-cost-explorer-to-analyze-data-transfer-costs/

These are the "Usage Type" that I tried to search for:

* `C3DataTransfer-Out-Bytes`: The amount of data transferred from Amazon S3 to Amazon EC2 within the same AWS Region
* `DataTransfer-Regional-Bytes`: The amount of data transferred from Amazon S3 to AWS resources within the same AWS Region

https://docs.aws.amazon.com/AmazonS3/latest/userguide/aws-usage-report-understand.html

----

From this post I understand that EC2 <-> S3 traffic is not charged:

> Bandwidth between EC2 and S3 is not charged (neither EC2 nor S3 will charge you, it's symmetric).

https://forums.aws.amazon.com/thread.jspa?threadID=11523

Reply:
yeah, this matches my understanding as well. It's why our s3 bill is ~$50/mo. Now, our bandwidth bill ... different story.